cr-examples/samples/src/main/java/oracle/code/samples/MathOptimizer.java

```diff
@@ -0,0 +1,281 @@
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package oracle.code.samples;
+
+import jdk.incubator.code.CodeReflection;
+import jdk.incubator.code.CopyContext;
+import jdk.incubator.code.Location;
+import jdk.incubator.code.Op;
+import jdk.incubator.code.OpTransformer;
+import jdk.incubator.code.TypeElement;
+import jdk.incubator.code.Value;
+import jdk.incubator.code.dialect.core.CoreOp;
+import jdk.incubator.code.dialect.java.JavaOp;
+import jdk.incubator.code.dialect.java.JavaType;
+import jdk.incubator.code.dialect.java.MethodRef;
+import jdk.incubator.code.interpreter.Interpreter;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * Simple example of how to use the code reflection API.
+ *
+ * <p>
+ * This example replaces a math function Math.pow with an optimized function using code transforms
+ * from the code-reflection API. The optimized function can be applied only under certain conditions.
+ * </p>
+ *
+ * <p>
+ * Optimizations:
+ * 1) Replace Pow(x, y) when x == 2 to (1 << y), if only if the parameter y is an integer.
+ * 2) Replace Pow(x, y) when y == 2 to (x * x).
+ * </p>
+ *
+ * <p>
+ *     Babylon repository: {@see <a href="https://github.com/openjdk/babylon/tree/code-reflection">link</a>}
+ * </p>
+ *
+ * <p>
+ *     How to run?
+ *     <code>
+ *         java --enable-preview -cp target/crsamples-1.0-SNAPSHOT.jar oracle.code.samples.MathOptimizer
+ *     </code>
+ * </p>:
+ */
+public class MathOptimizer {
+
+    @CodeReflection
+    private static double myFunction(int value) {
+        return Math.pow(2, value);
+    }
+
+    // if pow(2, x), then substitute for this function
+    // We could apply this function if, at runtime, user pass int values to the pow function
+    // Thus, we narrow the result type from 8 bytes (double) to 4 bytes (INT).
+    private static int functionShift(int val) {
+        return 1 << val;
+    }
+
+    // if pow(x, 2) then substitute for this function
+    private static double functionMult(double x) {
```

Comment:
A more complex example to consider later after this PR would be to make these methods reflectable, and then you can inline these methods into the transformed model so `Math.pow` is substituted with direct expressions.